xds/src/main/java/io/grpc/xds/ServerXdsClient.java

```diff
@@ -0,0 +1,237 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.envoyproxy.envoy.config.core.v3.Address;
+import io.envoyproxy.envoy.config.listener.v3.FilterChain;
+import io.envoyproxy.envoy.config.listener.v3.FilterChainMatch;
+import io.envoyproxy.envoy.config.listener.v3.Listener;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.xds.EnvoyProtoData.Node;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * XdsClient implementation for server side usages.
+ */
+final class ServerXdsClient extends AbstractXdsClient {
+
+  // Longest time to wait, since the subscription to some resource, for concluding its absence.
+  @VisibleForTesting
+  static final int INITIAL_RESOURCE_FETCH_TIMEOUT_SEC = 15;
+  private final SynchronizationContext syncContext;
+  @Nullable
+  private ListenerWatcher listenerWatcher;
+  private int listenerPort = -1;
+  @Nullable
+  private ScheduledHandle ldsRespTimer;
+
+  ServerXdsClient(
+      XdsChannel channel,
+      Node node,
+      SynchronizationContext syncContext,
+      ScheduledExecutorService timeService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Stopwatch stopwatch) {
+    super(channel, node, timeService, backoffPolicyProvider, stopwatch);
+    this.syncContext = checkNotNull(syncContext, "syncContext");
+  }
+
+  @Override
+  void watchListenerData(int port, ListenerWatcher watcher) {
+    checkState(listenerWatcher == null, "ListenerWatcher already registered");
+    listenerWatcher = checkNotNull(watcher, "watcher");
+    checkArgument(port > 0, "port needs to be > 0");
+    this.listenerPort = port;
+    getLogger().log(XdsLogLevel.INFO, "Started watching listener for port {0}", port);
+    updateNodeMetadataForListenerRequest(port);
+    adjustResourceSubscription(ResourceType.LDS);
+    ldsRespTimer =
```

Comment:
Few comments about this sequence of actions:

- if `adjustResourceSubscription` decided to not issue a request because it's in backoffMode do you still need to set the timer?
- why `updateNodeMetadataForListenerRequest` if you are in backoffMode and not going to issue the request?
- can the INFO log statement be issued only if a request is issued (and the timer set)?
- if the ServerXdsClient is going to be shared among multiple server instances then updating the common `node` instance is going to be problematic. Hopefully with the new server side API, modifying Node will not be needed. So we can ignore this problem for now 

I understand some of the above was existing code but I understand one of the goals is to fix whatever issues we can fix.

Reply:
- if `adjustResourceSubscription` decided to not issue a request because it's in backoffMode do you still need to set the timer?

Yeah, this is something missed. It should check before scheduling the timer. It should call `isInBackoff()` to check before scheduling the timer. Good catch. Fixed it.

- why `updateNodeMetadataForListenerRequest` if you are in backoffMode and not going to issue the request?

I think in the existing `XdsClientImpl2` this is wrong. It should indeed update the Node even if currently in backoff and no request is being sent. Currently in backoff just means we should not send the request now, the retry task will send the request once the timer fires. So even if the retry timer is pending, we should set up the Node properly so that once the timer fires it picks up the latest updated Node for sending the request.

- can the INFO log statement be issued only if a request is issued (and the timer set)?

I think it is ok as the log just says it *starts watching*, which means the watcher is added, not necessarily the resource is requested. It's just for business logic level (the caller of this API does not care about how XdsClient subscribe to the resources, it just need to know that its watcher is added and will be notified later). There are other logs for requests being sent.